### PR TITLE
Sd 566 extend queryservice with

### PR DESCRIPTION
**Pull request by filipdujmusic**

### CODE REVIEW

File: contracts/services/QueryService.sol [hunk 1]
```diff
@@ -363,4 +399,25 @@ contract QueryService is IVersioned {
...
+        // COUNT TOTAL RESPONSE SIZE
+        uint256 responseItemsCount = 0;
+        for (uint256 i = 0; i < assetFactories.length; i++) {
+            address[] memory instances = IAssetFactoryCommon(assetFactories[i]).getInstancesForIssuer(issuer);
+            for (uint256 j = 0; j < instances.length; j++) {
+                if (IToken(instances[j]).balanceOf(investor) > 0) { 
+                    responseItemsCount++;
+                }
+            }
+        }
```
Comment (mightymatth) [inline contracts/services/QueryService.sol:418]: You're iterating through factories and their assets:
* once (here) - to get array size, and
* again (in the main method) - to make the response.

Isn't it double work that can be done in one pass? Consider that you're double-checking investors' balance in every single pass.
If you need to know the array size for structs ahead of time, can't you just create a map (`mapping(uint => Structs.AssetBalance)` where the keys would be array indices (1,2,3...), and then, in the end, you will end up with the map and the total number for items in the response. Then you could just create an array, and fill it with the data in the loop over map items. Does it make sense? Or is your approach really more efficient to double iterate to save memory allocation frequency? We can discuss this in person if you like.

File: contracts/shared/Structs.sol [hunk 2]
```diff
@@ -252,6 +252,15 @@ contract Structs {
...
     }
+    
+    struct AssetBalance {
+        address contractAddress;
+        uint8 decimals;
+        string name;
+        string symbol;
+        uint256 balance;
+        AssetCommonState assetCommonState;
+    }
```
Comment (mightymatth) [inline contracts/shared/Structs.sol:263]: I like the response, so consider this as approved :)

File: contracts/services/QueryService.sol [hunk 3]
```diff
@@ -350,6 +350,42 @@ contract QueryService is IVersioned {
         return response; 
     }
 
+    function getAssetBalancesForIssuer(
+        address issuer,
+        address investor,
+        address[] memory assetFactories,
+        address[] memory campaignFactories
```
Comment (mightymatth) [inline contracts/services/QueryService.sol:357]: I'm not sure why there is `campaignFactories` if you don't use them. Aren't just `assetFactories` sufficient enough to list the assets in this phase?

---

When we talked about the issue where we would have campaigns that would sometimes have duplicate items in the response, couldn't we just fill the map (`mapping(uint => Structs.AssetBalance)` where the key would be the token address so we wouldn't have duplicates? 
I posted it here just as a reminder.

File: util/helpers.ts [hunk 4]
```diff
@@ -936,3 +996,34 @@ export async function queryCampaignsForIssuerInvestor(
...
+  issuer: string;
+}
+interface AssetBalance {
+  contractAddress: string;
+  decimals: number;
+  name: string;
+  symbol: string;
+  balance: BigNumber;
+  assetCommonState: AssetCommonState
+}
```
Comment (githinho) [inline util/helpers.ts:1029]: Add a new line.

File: contracts/services/QueryService.sol (hunk 3 above)
Comment (githinho) [reply contracts/services/QueryService.sol:357]: Add comment why `campaignFactories` is not used.

**Review by githinho — APPROVED** [review]

**Pull request merged**